ledger-service/http-json/src/main/scala/com/digitalasset/http/json/JsonProtocol.scala

```diff
@@ -348,12 +345,37 @@ object JsonProtocol extends DefaultJsonProtocol {
       }
     }
 
-  implicit val WarningsWrapperFormat: RootJsonFormat[domain.WarningsWrapper] =
-    jsonFormat1(domain.WarningsWrapper)
+  implicit val AsyncWarningsWrapperFormat: RootJsonFormat[domain.AsyncWarningsWrapper] =
+    jsonFormat1(domain.AsyncWarningsWrapper)
 
   implicit val UnknownTemplateIdsFormat: RootJsonFormat[domain.UnknownTemplateIds] = jsonFormat1(
     domain.UnknownTemplateIds)
 
   implicit val UnknownPartiesFormat: RootJsonFormat[domain.UnknownParties] = jsonFormat1(
     domain.UnknownParties)
+
+  implicit def OkResponseFormat[R: JsonFormat]: RootJsonFormat[domain.OkResponse[R]] =
+    jsonFormat3(domain.OkResponse[R])
+
+  implicit val ErrorResponseFormat: RootJsonFormat[domain.ErrorResponse] =
+    jsonFormat3(domain.ErrorResponse)
+
+  implicit def SyncResponseFormat[R: JsonFormat]: RootJsonFormat[domain.SyncResponse[R]] =
+    new RootJsonFormat[domain.SyncResponse[R]] {
+      private val resultKey = "result"
+      private val errorsKey = "errors"
+
+      override def write(obj: domain.SyncResponse[R]): JsValue = obj match {
+        case a: domain.OkResponse[_] => OkResponseFormat[R].write(a)
+        case b: domain.ErrorResponse => ErrorResponseFormat.write(b)
+      }
+
+      override def read(json: JsValue): domain.SyncResponse[R] = json match {
+        case JsObject(fields) if fields.contains(resultKey) => OkResponseFormat[R].read(json)
+        case JsObject(fields) if fields.contains(errorsKey) => ErrorResponseFormat.read(json)
```

Comment:
If we do need to decode them, best avoid ambiguity:

```suggestion
        case JsObject(fields) =>
          (fields get resultKey, fields get errorKey) match {
            case (Some(result), None) => OkResponseFormat[R].read(json)
            case (None, Some(errors)) => ErrorResponseFormat.read(json)
            // error cases
```